pkg/controller/servicemesh/memberroll/controller_test.go

```diff
@@ -464,6 +464,33 @@ func TestReconcileNamespacesIgnoresControlPlaneNamespace(t *testing.T) {
 	assert.DeepEquals(configuredMembers, []string{appNamespace}, "reconcileNamespaces returned an unexpected configuredMembers list", t)
 }
 
+func TestReconcileNamespacesWorksWithMultipleNamespaces(t *testing.T) {
+	namespace := newAppNamespace(appNamespace)
+	namespace2 := newAppNamespace(appNamespace2)
+
+	_, _, r, nsReconciler, _ := createClientAndReconciler(t, namespace, namespace2)
+
+	reqLogger := log.WithValues("ServiceMeshMemberRoll", request)
+
+	namespaces := sets.NewString(controlPlaneNamespace, appNamespace, appNamespace2)
+	configuredMembers, err, nsErrors := r.reconcileNamespaces(
+		namespaces,
+		sets.String{},
+		controlPlaneNamespace,
+		meshVersionDefault,
+		reqLogger,
+	)
+	if err != nil {
+		t.Fatalf("reconcileNamespaces failed: %v", err)
+	}
+	if len(nsErrors) > 0 {
+		t.Fatalf("reconcileNamespaces returned unexpected nsErrors: %v", nsErrors)
+	}
+
+	assertNamespaceReconcilerInvoked(t, nsReconciler, appNamespace, appNamespace2) // NOTE: no controlPlaneNamespace
+	assert.DeepEquals(configuredMembers, []string{appNamespace, appNamespace2}, "reconcileNamespaces returned an unexpected configuredMembers list", t)
+}
```

Comment:
If I revert the changes in controller.go, this test doesn't fail. The key change seems to be in `Reconcile()`, so I'd expect the test to call it instead of `reconcileNamespaces`. What was the exact cause of the infinite loop? I can't figure it out just by looking at the log/code.

Reply:
You're right. I pushed a weird mix of two different approaches to fixing the problem; first I thought  I could fix it just in this function, but later I realized that the way we're using the return value from it is also wrong. I pushed the correct test